### DISCUSSION TITLE
Task with retries hangs in Pytest (Celery@5.4.0 with redis and pytest-celery < 1.0.0)

### DISCUSSION THREAD

**Discussion by gnottobfly** [celery/celery#9076] [Q&A] [answered]
Hi folks,

I found that when running multiple tests in pytest, if a task encounters a retry then that task will timeout. I've created a repository with the reproduction steps using the minimal number of code/dependencies. See it here: https://github.com/gnottobfly/celery-worker-pytest-retry-issue.

If I run the test one at a time, they all pass. But when running the whole suite together, the retry task fails. I'd like to be able to mock out dependencies for integration tests which is why I want to avoid using the new plugin or using a docker compose setup.

Has anyone encountered this before?

**Comment by gnottobfly** [comment]
Just some more context, I see in Redis the task information here:
```
{"status": "RETRY", "result": {"exc_type": "Exception", "exc_message": ["Boom!"], "exc_module": "builtins"}, "traceback": "Traceback (most recent call last):\n  File \"/Users/gnotto/sandbox/celery-worker-pytest-retry-issue/test.py\", line 37, in work_it\n    do_work()\n  File \"/Users/gnotto/.pyenv/versions/3.11.8/lib/python3.11/unittest/mock.py\", line 1124, in __call__\n    return self._mock_call(*args, **kwargs)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/gnotto/.pyenv/versions/3.11.8/lib/python3.11/unittest/mock.py\", line 1128, in _mock_call\n    return self._execute_mock_call(*args, **kwargs)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/gnotto/.pyenv/versions/3.11.8/lib/python3.11/unittest/mock.py\", line 1187, in _execute_mock_call\n    raise result\nException: Boom!\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/Users/gnotto/.pyenv/versions/3.11.8/envs/sandbox/lib/python3.11/site-packages/celery/app/trace.py\", line 453, in trace_task\n    R = retval = fun(*args, **kwargs)\n                 ^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/gnotto/sandbox/celery-worker-pytest-retry-issue/test.py\", line 39, in work_it\n    raise task.retry(exc=e)\n          ^^^^^^^^^^^^^^^^^\n  File \"/Users/gnotto/.pyenv/versions/3.11.8/envs/sandbox/lib/python3.11/site-packages/celery/app/task.py\", line 757, in retry\n    raise ret\ncelery.exceptions.Retry: Retry in 1s: Exception('Boom!')\n", "children": [[["cc0035db-2433-4bc9-8ea2-5069e8521216", null], null]], "date_done": null, "task_id": "cc0035db-2433-4bc9-8ea2-5069e8521216"}
```

The issue is that the worker doesn't seem to be picking up the task.

**Comment by Nusnus** [comment, marked as answer]
This issue also happens with the latest pytest-celery due to a bug with Kombu & Redis.

There's already a fix cooking up - https://github.com/celery/kombu/pull/2007

When tests get to D/C from Redis and the worker is expected to reconnect, the bug is reproduced and the worker hangs. We will release the fix in the next release of Kombu and Celery to fix the issue from its core.

For now, using Kombu 5.2.2 should avoid this situation which was introduced since Kombu 5.2.3.

**Reply by gnottobfly** [reply]
Just want to check, does this bug also affect production code using real celery workers?

**Reply by Nusnus** [reply]
> Just want to check, does this bug also affect production code using real celery workers?

Unfortunately, yes.
Using Kombu >= 5.2.3 in production will hang the Celery worker after D/C with a Redis broker.

The expected fix will be released with Kombu 5.4.

**Comment by gnottobfly** [comment]
Really appreciate the quick reply here! 🙌

**Reply by Nusnus** [reply]
♥️♥️
